### ISSUE TITLE
GenevaMetricExporter: Add ResourceBuilder support

### ISSUE THREAD

**Issue by dpk83** [open-telemetry/opentelemetry-dotnet-contrib#835]
# Issue with OpenTelemetry.Exporter.Geneva

**Is this a feature request or a bug?**

* [x ] Feature Request
* [ ] Bug

**What is the expected behavior?**

GenevaMetricExporter should respect the resource builder and add attributes from resource to the emitted metrics.

What do you expect to see?

As a user when I set a resource builder the attributes detected from resources are added as dimensions(aka tags) to all metrics exported to Geneva.

**What is the actual behavior?**
Currently resource builder is completely ignored by Geneva metric exporter.

## Additional Context

Add any other context about the feature request here.

**Comment by reyang** [comment]
I haven't spent time thinking through this, my gut feeling is that there will be a combination of the following two features and I believe this issue is more about the 2nd one:
1. In the OpenTelemetry SDK, the [View API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/metrics/sdk.md#view) will allow the application/service developer to configure a particular metrics stream to use certain dimensions from Resource.
2. In the Geneva Exporter, there is a mechanism which allows Resources (all or user specified) to be converted to dimensions.

I think 2 should be fairly easy as long as we decide the behavior. There is one design decision we'll have to make - will resources be used by default as dimensions or not? (e.g. I don't think it's good idea to take all Resources and convert them to dimensions by default).

**Comment by cijothomas** [comment]
1 -> The OTel SDK's View does not have the capability to use Resource as the dimension. 

2 is reasonable, with opt-in behavior, and probably another flag to determine the conflict-resolution (does Resource override samenamed attribute that is provided during reporting of measurement.?

**Comment by cizra** [comment]
It is also possible to attach attributes to Meters, and instruments like Histogram. These attributes also don't get picked up by Geneva metric exporter. Would adding support for these fit into this feature request, or should I file a new one?

**Comment by cijothomas** [comment]
> It is also possible to attach attributes to Meters, and instruments like Histogram. These attributes also don't get picked up by Geneva metric exporter. Would adding support for these fit into this feature request, or should I file a new one?

The upcoming version of GenevaExporter will be using otlp proto format as its payload, so anything supported in OTLP would be automatically supported. That includes Resource and Meter attributes. 
(Note: Instrument level attributes is not really a spec thing, it exists in the API, but not supported in the OTel SDK itself. See https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics#metrics-enrichment)

**Comment by cizra** [comment]
Very interesting, thanks @cijothomas . Does it imply that at some point, GenevaExporter could be deprecated and expect Geneva MA would accept [OtlpExporter](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol)?

**Comment by cijothomas** [comment]
> Very interesting, thanks @cijothomas . Does it imply that at some point, GenevaExporter could be deprecated and expect Geneva MA would accept [OtlpExporter](https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol)?

No. GenevaExporter is designed to export to operating system native systems (ETW in Windows, user_events in Linux soon), and that is required for high performance scenarios.

**Comment by vishweshbankwar** [comment]
> > It is also possible to attach attributes to Meters, and instruments like Histogram. These attributes also don't get picked up by Geneva metric exporter. Would adding support for these fit into this feature request, or should I file a new one?
> 
> The upcoming version of GenevaExporter will be using otlp proto format as its payload, so anything supported in OTLP would be automatically supported. That includes Resource and Meter attributes. (Note: Instrument level attributes is not really a spec thing, it exists in the API, but not supported in the OTel SDK itself. See https://github.com/open-telemetry/opentelemetry-dotnet/tree/main/docs/metrics#metrics-enrichment)

This is now available with https://www.nuget.org/packages/OpenTelemetry.Exporter.Geneva/1.8.0 as an experimental feature. See [instructions](https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Exporter.Geneva#:~:text=Set%20PrivatePreviewEnableOtlpProtobufEncoding%3Dtrue%20to%20opt%2Din%20to%20the%20experimental%20feature%20for%20changing%20the%20underlying%20serialization%20format%20to%20binary%20protobuf%20following%20the%20schema%20defined%20in%20OTLP%20specification.) to onboard.

**Comment by DavidParks8** [comment]
@vishweshbankwar, when will the experimental feature be available for linux?

**Comment by vishweshbankwar** [comment]
@DavidParks8 - Aug/Sept is tentative timeframe.

**Comment by cataggar** [comment]
I was expecting `ResourceBuilder` would set resource attributes like `PrepopulatedFields`. Is that the wrong expectation?  Please see this [internal PR comment](https://dev.azure.com/msazure/One/_git/AzS-Conveyor-Telemetry/pullRequest/12059704#1743021265).

**Comment by cataggar** [comment]
I was expecting `ResourceBuilder` to set resource attributes like `PrepopulatedFields` & `PrepopulatedMetricDimensions`. Is that the wrong expectation? I tried this out yesterday in an internal PR with:

```xml
<PackageVersion Include="OpenTelemetry.Exporter.Geneva" Version="1.11.1" />
```

And some relevant code snippets:
```fsharp
        logging.AddOpenTelemetry(fun otelLogging ->
            otelLogging.SetResourceBuilder mdsdResourceBuilder |> ignore


        openTelemetry.WithTracing(fun tracing ->
            tracing.SetResourceBuilder mdsdResourceBuilder |> ignore


        openTelemetry.WithMetrics(fun metrics ->
            metrics.SetResourceBuilder mdmResourceBuilder |> ignore
```

Since that did not work as expected, snippets from my workaround:
```fsharp
    let mdmResourceBuilder = telemetryConfig.MdmResourceBuilder |> Option.defaultValue (createMdmResourceBuilder telemetryConfig.GenevaMetrics)
    let mdsdResourceBuilder = telemetryConfig.MdsdResourceBuilder |> Option.defaultValue (createMdsdResourceBuilder telemetryConfig.GenevaMetrics)


options.PrepopulatedFields <- mdsdResourceAttributes


genevaMetrics.PrepopulatedMetricDimensions <- mdmResourceAttributes
```

**Comment by cijothomas** [comment]
> I was expecting ResourceBuilder to set resource attributes like PrepopulatedFields & PrepopulatedMetricDimensions. Is that the wrong expectation?

Yes that is the wrong expectation!
For metrics, if you use the https://github.com/open-telemetry/opentelemetry-dotnet-contrib/tree/main/src/OpenTelemetry.Exporter.Geneva#otlpprotobufencoding preview feature, Resource is serialized and sent to Geneva.

For logs,traces - prepopulatedfields is the only current option. Resource support may come in the future, but not firm timelines.

**Comment by sandersaares** [comment]
I would also expect resources to simply set (until proper support exists) the prepopulated fields. I understand on the wire it should ideally not just replicate like that but why would this be the wrong expectation from a UX perspective?

**Comment by cijothomas** [comment]
> I would also expect resources to simply set (until proper support exists) the prepopulated fields. I understand on the wire it should ideally not just replicate like that but why would this be the wrong expectation from a UX perspective?

Not clear what is the ask.. What is the UX being referred to here?

**Comment by sandersaares** [comment]
The way I see it as a user: a resource is a way to attach (reused) metadata to telemetry items. While in an ideal world, it would only be transmitted once and then referenced logically, if the support for doing this "the proper way" is lacking, this metadata needs to be attached individually to each item (metric, log entry, trace entry) by the SDK.

**Comment by cijothomas** [comment]
> this metadata needs to be attached individually to each item (metric, log entry, trace entry) by the SDK.

Yes. that is correct. ETW/user-events don't have the native notion of Resource like things - so everything is send with each telemetry item.

**Comment by sandersaares** [comment]
Just to emphasize, "by the SDK" is the operative term in my description - as a user, I expect the SDK (into which I categorize the exporter, as well) to do this automatically if it cannot attach resources the "proper" way.

**Comment by cijothomas** [comment]
> Just to emphasize, "by the SDK" is the operative term in my description - as a user, I expect the SDK (into which I categorize the exporter, as well) to do this automatically if it cannot attach resources the "proper" way.

Got it. It won't be automatic for ETW (and Linux tracepoints too) - it'll be opt-in. i.e app owner who is setting up the Exporter has to tell which resource attributes should be attached to each telemetry item before writing to ETW/Tracepoints. This is very similar to OTel Rust's ETW/user-events exporters work : https://github.com/open-telemetry/opentelemetry-rust-contrib/blob/main/opentelemetry-user-events-logs/src/lib.rs#L81-L92

IIUC, that capability is already part of https://github.com/open-telemetry/opentelemetry-dotnet-contrib/blob/main/src/OpenTelemetry.Exporter.Geneva/CHANGELOG.md#1150 

Please check and reply if that is what you were looking for.

**Comment by sandersaares** [comment]
My expectation is that this is automatic. In that sense an opt-in mechanism does not meet my expectations. Functionally, of course, it's better than nothing but it does not make me a happy customer to have to do any manual labor to wire up what should (in my mind) be wired up automatically. So it works in a pinch but it leaves a sour taste in the mouth.

**Comment by cijothomas** [comment]
> My expectation is that this is automatic. In that sense an opt-in mechanism does not meet my expectations. Functionally, of course, it's better than nothing but it does not make me a happy customer to have to do any manual labor to wire up what should (in my mind) be wired up automatically. So it works in a pinch but it leaves a sour taste in the mouth.

Got it. I don't think we can do automatic without major version bump (Also, I don't agree that all resource attributes should be automatically attached to each event when it comes to ETW- but I am not an owner of this component anymore) If we do that, it'll be a breaking behavior change, and we may end up with too many resource attributes that may not be what user had intended. (Note that Resource is shared across exporters, so there are users who does more than one exporter and for OTLP like things, Resource is probably fine, as it is efficiently represented, but for Kernel primitives - we are writing it with each event, so perf will suffer)

Selectively opting into Resource attributes at exporter/ingestion/backend level is quite common in OTel world from what I've seen. Prometheus's native OTLP ingestion also had this explicit opt-in - users had to explicitly state which Resoure attributes they want stored. Everything else got dropped by default. Things may have changed recently in Prometheus with `target_info`. And IIRC, this was standard for Zipkin/Jaeger too at one point.
